### PR TITLE
chore(NODE-5829): update driver dependencies

### PR DESCRIPTION
**Pull request by aditi-khare-mongoDB**
### Description
Update dev, prod, and peer dependencies in the driver to latest safe and compatible versions + ones with minimal code changes.

##### Dependency Update
No Peer Dependency Updates until next major release

Dev Dependency Updates
```
 @aws-sdk/credential-providers  ^3.465.0  →   ^3.515.0
 @microsoft/api-extractor        ^7.38.4  →    ^7.40.6
 @octokit/core                    ^4.2.4  →     ^5.1.0
 @types/node                    ^20.10.3  →  ^20.11.20
 @types/semver                    ^7.5.6  →     ^7.5.7
 @types/whatwg-url               ^11.0.3  →    ^11.0.4
 eslint                          ^8.55.0  →    ^8.56.0
 eslint-plugin-import            ^2.29.0  →    ^2.29.1
 mocha                           ^10.2.0  →    ^10.3.0
 semver                           ^7.5.4  →     ^7.6.0
 socks                            ^2.7.1  →     ^2.8.1
 ts-node                         ^10.9.1  →    ^10.9.2
 tsd                             ^0.28.1  →    ^0.30.6
 chai                            ^4.3.10  →     ^4.4.1
```

Production Dependency Updates
n/a at this time 

Note: `sinon` and `@types/sinon` will be updated after node v16 bug fix is done in [NODE-5942](https://jira.mongodb.org/browse/NODE-5942)

##### Is there new documentation needed for these changes?
No.

#### What is the motivation for this change?
QOL

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

**Comment by baileympearson** [comment]
here's what I see locally:
```
Package                           Current   Wanted  Latest  Location                                       Depended by
@microsoft/api-extractor           7.38.5   7.40.5  7.40.5  node_modules/@microsoft/api-extractor          node-mongodb-native
@types/sinon                      10.0.20  10.0.20  17.0.3  node_modules/@types/sinon                      node-mongodb-native
@typescript-eslint/eslint-plugin   5.62.0   5.62.0   7.0.2  node_modules/@typescript-eslint/eslint-plugin  node-mongodb-native
@typescript-eslint/parser          5.62.0   5.62.0   7.0.2  node_modules/@typescript-eslint/parser         node-mongodb-native
chai                               4.3.10    4.4.1   5.1.0  node_modules/chai                              node-mongodb-native
chalk                               4.1.2    4.1.2   5.3.0  node_modules/chalk                             node-mongodb-native
eslint                             8.55.0   8.56.0  8.56.0  node_modules/eslint                            node-mongodb-native
eslint-config-prettier             8.10.0   8.10.0   9.1.0  node_modules/eslint-config-prettier            node-mongodb-native
eslint-plugin-import               2.29.0   2.29.1  2.29.1  node_modules/eslint-plugin-import              node-mongodb-native
eslint-plugin-prettier              4.2.1    4.2.1   5.1.3  node_modules/eslint-plugin-prettier            node-mongodb-native
eslint-plugin-simple-import-sort   10.0.0   10.0.0  12.0.0  node_modules/eslint-plugin-simple-import-sort  node-mongodb-native
eslint-plugin-unused-imports        2.0.0    2.0.0   3.1.0  node_modules/eslint-plugin-unused-imports      node-mongodb-native
gcp-metadata                        5.3.0    5.3.0   6.1.0  node_modules/gcp-metadata                      node-mongodb-native
kerberos                          MISSING    2.1.0   2.1.0  -                                              node-mongodb-native
prettier                            2.8.8    2.8.8   3.2.5  node_modules/prettier                          node-mongodb-native
sinon                              15.2.0   15.2.0  17.0.1  node_modules/sinon                             node-mongodb-native
ts-node                            10.9.1   10.9.2  10.9.2  node_modules/ts-node                           node-mongodb-native
tsd                                0.28.1   0.28.1  0.30.5  node_modules/tsd                               node-mongodb-native
typescript                          5.0.4    5.3.3   5.3.3  node_modules/typescript                        node-mongodb-native
```

- sinon its types types are okay to leave for now, we have https://jira.mongodb.org/browse/NODE-5942
- chalk I think is esmodule only now
- chai is esmodule only now
- typescript can be left alone

ts-node, tsd, api-extractor can probably be upgraded (as per discussions on other PRs)

1. what about all the eslint plugins?  
2. did you mean to update `@aws-sdk/credential-providers`? the PR description says no peer dependencies were updated

**Comment by aditi-khare-mongoDB** [comment]
@baileympearson 

I did mean to update the `aws-sdk` peer dependency, [per this slack conversation](https://mongodb.slack.com/archives/GGWBN4ZNK/p1708027707392869?thread_ts=1708025612.915399&cid=GGWBN4ZNK) (i've added a clarification in the PR description). 

The `eslint-prettier` updates cause import errors on my end.

File: src/cmap/auth/scram.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { saslprep } from '@mongodb-js/saslprep';
+import saslprep = require('@mongodb-js/saslprep');
```
Comment (nbbeeken) [inline src/cmap/auth/scram.ts:1]: While this seems minimal this is not a desirable change, this syntax is for supporting legacy modules that declare their types as `export =` and it changes the javascript generated at runtime:


Before this change:
```ts
const saslprep_1 = require("@mongodb-js/saslprep");
// ...
(0, saslprep_1.saslprep)(password)
```

With this change:
```ts
const saslprep = require("@mongodb-js/saslprep");
// ...
saslprep(password)
```


IIRC using the saslprep's default (CJS default) export as a function breaks bundlers. 

Let's file a ticket to fix the typescript on the saslprep package and leave it on the version it is currently at for now.

File: package.json [hunk 2]
```diff
@@ -36,7 +36,7 @@
     "kerberos": "^2.0.1",
     "mongodb-client-encryption": ">=6.0.0 <7",
     "snappy": "^7.2.2",
-    "socks": "^2.7.1"
+    "socks": "^2.7.3"
```
Comment (nbbeeken) [inline package.json:39]: This is a change to the peer deps, we can only change it in [dev deps](https://mongodb.slack.com/archives/GGWBN4ZNK/p1708028241681019?thread_ts=1708025612.915399&cid=GGWBN4ZNK).

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/cmap/auth/scram.ts (hunk 1 above)
Comment (nbbeeken) [reply src/cmap/auth/scram.ts:1]: https://jira.mongodb.org/browse/NODE-5963

**Comment by nbbeeken** [comment]
@aditi-khare-mongoDB Let me know if anything is blocking the changes I requested

**Comment by aditi-khare-mongoDB** [comment]
@baileympearson @nbbeeken 

You're right no peer updates, the updated dependencies are the following (only dev dependencies, no prod easily upgradable atm), and I've deleted my duplicate follow-up ticket since Neal's filed NODE-5693 already.

```
@aws-sdk/credential-providers  ^3.465.0  →   ^3.515.0
@microsoft/api-extractor        ^7.38.4  →    ^7.40.6
@octokit/core                    ^4.2.4  →     ^5.1.0
@types/node                    ^20.10.3  →  ^20.11.20
@types/semver                    ^7.5.6  →     ^7.5.7
@types/whatwg-url               ^11.0.3  →    ^11.0.4
eslint                          ^8.55.0  →    ^8.56.0
eslint-plugin-import            ^2.29.0  →    ^2.29.1
mocha                           ^10.2.0  →    ^10.3.0
semver                           ^7.5.4  →     ^7.6.0
socks                            ^2.7.1  →     ^2.8.1
ts-node                         ^10.9.1  →    ^10.9.2
tsd                             ^0.28.1  →    ^0.30.6
 ```

File: src/cmap/auth/scram.ts (hunk 1 above)
Comment (baileympearson) [reply src/cmap/auth/scram.ts:1]: just to confirm - we want to revert these changes too, right?

Comment (nbbeeken) [reply src/cmap/auth/scram.ts:1]: Yep, re-request for viz, we do not want this change

File: package-lock.json [hunk 3]
```diff
@@ -1549,9 +1657,9 @@
       }
     },
     "node_modules/@mongodb-js/saslprep": {
-      "version": "1.1.0",
-      "resolved": "https://registry.npmjs.org/@mongodb-js/saslprep/-/saslprep-1.1.0.tgz",
-      "integrity": "sha512-Xfijy7HvfzzqiOAhAepF4SGN5e9leLkMvg/OPOF97XemjfVCYN/oWa75wnkc6mltMSTwY+XlbhWgUOJmkFspSw==",
+      "version": "1.1.4",
```
Comment (nbbeeken) [inline package-lock.json:1660]: Saslprep is still upgraded, this should go back to the version it is on main. 

To be sure there aren't other accidental updates can you: 

- revert package-lock.json to be the same as main 
- rm -rf node_modules, 
- run npm install

then the `package-lock.json` should be upgraded to what we've modified in the package.json

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: package-lock.json (hunk 3 above)
Comment (aditi-khare-mongoDB) [reply package-lock.json:1660]: Ah I see, I was missing the `rm -rf node_modules`, thanks for the reminder!

**Review by nbbeeken — APPROVED** [review]

**Pull request merged**